src/java.desktop/share/classes/sun/java2d/loops/GraphicsPrimitiveMgr.java

```diff
@@ -274,6 +241,47 @@ private static void writeLog(String str) {
         }
     }
 
+    /**
+     * A holder for general primitives to avoid circular dependencies
+     * between GraphicsPrimitiveMgr and Blit/etc classes.
+     */
+    final static class GeneralPrimitives {
+
+        private static GraphicsPrimitive[] primitives;
+
+        /**
+         * Registers the general loop which will be used to produce specific
+         * primitives by the {@link GraphicsPrimitive#makePrimitive} function.
+         *
+         * @param gen the graphics primitive to be registered as the general loop
+         */
+        static synchronized void register(GraphicsPrimitive gen) {
+            if (primitives == null) {
+                primitives = new GraphicsPrimitive[]{gen};
+                return;
+            }
+            int len = primitives.length;
+            GraphicsPrimitive[] newGen = new GraphicsPrimitive[len + 1];
+            System.arraycopy(primitives, 0, newGen, 0, len);
+            newGen[len] = gen;
+            primitives = newGen;
+        }
+
+        static GraphicsPrimitive locate(int primTypeID) {
```

Comment:
it should be marked as `synchronized` too. Otherwise there will possible data race between other thread called `register`

Reply:
Fixed, thanks